xds/internal/balancer/priority/balancer_priority.go

```diff
@@ -0,0 +1,360 @@
+			continue
+		}
+
+		if !child.started ||
+			child.state.ConnectivityState == connectivity.Ready ||
+			p == len(pb.priorities)-1 {
+			if pb.childInUse != "" && pb.childInUse != child.name {
+				// childInUse was set and is different from this child, will
+				// change childInUse later. We need to update picker here
+				// immediately so parent stops using the old picker.
+				pb.cc.UpdateState(child.state)
+			}
+			pb.logger.Warningf("switching to (%q, %v) in syncPriority", child.name, p)
+			pb.switchToChild(child, p)
+			child.sendUpdate()
+			break
+		}
+	}
+}
+
+// Stop priorities [p+1, lowest].
+//
+// caller must hold pb.mu.
+func (pb *priorityBalancer) stopSubBalancersLowerThanPriority(p int) {
+	for i := p + 1; i < len(pb.priorities); i++ {
+		name := pb.priorities[i]
+		child, ok := pb.children[name]
+		if !ok {
+			pb.logger.Warningf("child with name %q is not found in children", name)
+			continue
+		}
+		child.stop()
+	}
+}
+
+// switchToChild does the following:
+// - stop all child with lower priorities
+// - if childInUse is not this child
+//   - set childInUse to this child
+//   - stops init timer
+//   - if this child is not started, start it, and start a init timer
+//
+// Note that it does NOT send the current child state (picker) to the parent
+// ClientConn. The caller needs to send it if necessary.
+//
+// this can be called when
+// 1. first update, start p0
+// 2. an update moves a READY child from a lower priority to higher
+// 2. a different builder is updated for this child
+// 3. a high priority goes Failure, start next
+// 4. a high priority init timeout, start next
+//
+// caller must hold pb.mu.
+func (pb *priorityBalancer) switchToChild(child *childBalancer, priority int) {
+	// Stop lower priorities even if childInUse is same as this child. It's
+	// possible this child was moved from a priority to another.
+	pb.stopSubBalancersLowerThanPriority(priority)
+
+	// If this child is already in use, do nothing.
+	//
+	// This can happen:
+	// - all priorities are not READY, an config update always triggers switch
+	// to the lowest. In this case, the lowest child could still be connecting,
+	// so we don't stop the init timer.
+	// - a high priority is READY, an config update always triggers switch to
+	// it.
+	if pb.childInUse == child.name && child.started {
+		return
+	}
+	pb.childInUse = child.name
+	pb.priorityInUse = priority
+
+	// Init timer is always for childInUse. Since we are switching to a
+	// different child, we will stop the init timer no matter what. If this
+	// child is not started, we will start the init timer later.
+	if pb.priorityInitTimer != nil {
+		pb.priorityInitTimer.Stop()
+		pb.priorityInitTimer = nil
+	}
+
+	if !child.started {
+		child.start()
+		pb.priorityInitTimer = time.AfterFunc(defaultPriorityInitTimeout, func() {
+			pb.mu.Lock()
+			defer pb.mu.Unlock()
+			if pb.childInUse != child.name {
+				return
+			}
+			pb.priorityInitTimer = nil
```

Comment:
This doesn't seem to be strictly necessary AFAICT.

Is it possible it causes problems, actually?  What if a new timer is created between this timer firing and `pb.mu.Lock()` unblocking (pretend there is a long sleep on the first line of the timer function)?  That would lead to the new timer being "lost", but still ticking.